Database connection control options

trueqbit opened the pull request:
* Introduced a connection control struct as an aggregate of options. This opens the way for various options and modes to establish a database connection, and can be used to make certain assumption in a multi-threading environment. E.g. `make_storage("", connection_control{.open_forever = true});`. Also see PR #1393.
* An 'on open' callback can now be specified in a declarative way to the `make_storage()` call. This finally solves the issue that in-memory databases didn't call the 'on open' callback. E.g. `make_storage("", on_open([](sqlite3* db) {}));`.
* Reordered members of the connection holder. This will be important for upcoming multi-threading improvements (see PR #1395), avoiding false L1 cache sharing.

trueqbit committed:
Database connection control options

* Introduced a connection control struct as an aggregate of options. This opens the way for various options and modes to establish a database connection.
* An 'on open' callback can now be specified in a declarative way to the `make_storage()` call. This finally solves the issue that in-memory databases didn't call the 'on open' callback.
* Reordered members of the connection holder. This will be important for upcoming multi-threading improvements, avoiding false L1 cache sharing.

trueqbit committed:
Updated amalgamated header

fnc12 commented on dev/connection_holder.h:
```diff
@@ -12,25 +13,38 @@ namespace sqlite_orm {
 
         struct connection_holder {
-            connection_holder(std::string filename) : filename(std::move(filename)) {}
+            connection_holder(std::string filename, std::function<void(sqlite3*)> onAfterOpen) :
```
nit: if we want to specify whether callback is going to be called before or after some event then we'd better avoid 'on*' naming and use 'willOpen' and 'didOpen' instead just like Apple does cause it is more concise than 'onAfterOpen' or 'onBeforeOpen'.

trueqbit replied:
Interesting 👍

fnc12 commented on tests/storage_tests.cpp:
```diff
@@ -4,6 +4,49 @@
+TEST_CASE("connection control") {
+    const auto openForever = GENERATE(false, true);
+    SECTION("") {
+        bool onOpenCalled = false;
```
I'd change `bool` to `int` to cover amount of `onOpen` calls count just in case. I know your code doesn't produce such a bug but if we cover it with small amount of energy then it will be a win

trueqbit replied:
thx! You just saved my butt :)

fnc12 commented on tests/storage_tests.cpp:
```diff
@@ -4,6 +4,49 @@
 
+#ifdef SQLITE_ORM_CTAD_SUPPORTED
+TEST_CASE("connection control") {
+    const auto openForever = GENERATE(false, true);
```
`GENERATE` solves the issue with parallels sections tree. I have to use it as well

trueqbit committed:
Avoided a triple-layered alias template expressions

Older compilers such as Visual C++ 16.11 [msvc 1929] choke on deeply nested alias template expressions and can be made happy by unraveling these expressions.

trueqbit committed:
Corrected the point at which the handler is called to set up the database

trueqbit committed:
Renamed database callback function variable

trueqbit committed:
Merge branch 'upstream/circumvent_triple_layered_alias_template' into upstream/connection_control
